core/transaction-graph/src/ready.rs

```diff
@@ -0,0 +1,554 @@
+		assert!(!self.ready.contains_key(&tx.transaction.hash), "Transaction is already imported.");
+
+		self.insertion_id += 1;
+		let insertion_id = self.insertion_id;
+		let hash = tx.transaction.hash.clone();
+		let tx = tx.transaction;
+
+		let replaced = self.replace_previous(&tx)?;
+
+		let mut goes_to_best = true;
+		// Add links to transactions that unlock the current one
+		for tag in &tx.requires {
+			// Check if the transaction that satisfies the tag is still in the queue.
+			if let Some(other) = self.provided_tags.get(tag) {
+				let mut tx = self.ready.get_mut(other).expect(HASH_READY);
+				tx.unlocks.push(hash.clone());
+				// this transaction depends on some other, so it doesn't go to best directly.
+				goes_to_best = false;
+			}
+	 	}
+
+		// update provided_tags
+		for tag in tx.provides.clone() {
+			self.provided_tags.insert(tag, hash.clone());
+		}
+
+		let transaction = TransactionRef {
+			insertion_id,
+			valid_till: block_number + tx.longevity,
+			transaction: Arc::new(tx),
+		};
+
+		// insert to best if it doesn't require any other transaction to be included before it
+		if goes_to_best {
+			self.best.insert(transaction.clone());
+		}
+
+		// insert to Ready
+		self.ready.insert(hash, ReadyTx {
+			transaction,
+			unlocks: vec![],
+			requires_offset: 0,
+		});
+
+		Ok(replaced)
+	}
+
+	/// Returns true if given hash is part of the queue.
+	pub fn contains(&self, hash: &Hash) -> bool {
+		self.ready.contains_key(hash)
+	}
+
+	/// Removes invalid transactions from the ready pool.
+	///
+	/// NOTE removing a transaction will also cause a removal of all transactions that depend on that one
+	/// (i.e. the entire subgraph that this transaction is a start of will be removed).
+	/// All removed transactions are returned.
+	pub fn remove_invalid(&mut self, hashes: &[Hash]) -> Vec<Arc<Transaction<Hash>>> {
+		let mut removed = vec![];
+		let mut to_remove = hashes.iter().cloned().collect::<Vec<_>>();
+
+		loop {
+			let hash = match to_remove.pop() {
+				Some(hash) => hash,
+				None => return removed,
+			};
+
+			if let Some(mut tx) = self.ready.remove(&hash) {
+				// remove entries from provided_tags
+				for tag in &tx.transaction.transaction.provides {
+					self.provided_tags.remove(tag);
+				}
+				// remove from unlocks
+				for tag in &tx.transaction.transaction.requires {
+					if let Some(hash) = self.provided_tags.get(tag) {
+						if let Some(tx) = self.ready.get_mut(hash) {
+							remove_item(&mut tx.unlocks, &hash);
+						}
+					}
+				}
+
+				// remove from best
+				self.best.remove(&tx.transaction);
+
+				// remove all transactions that the current one unlocks
+				to_remove.append(&mut tx.unlocks);
+
+				// add to removed
+				debug!(target: "txpool", "[{:?}] Removed as invalid: ", hash);
+				removed.push(tx.transaction.transaction);
+			}
+		}
+	}
+
+	/// Removes transactions that provide given tag.
+	///
+	/// All transactions that lead to a transaction, which provides this tag
```

Comment:
So if `C` can now be pruned it means that either:
1. All requirements for `C` are now satisfied, meaning both `A->B` and `D->E` chains (or some replacement chain) had to be included as well.
2. Some other transaction provided `C` and it got finalized, which doesn't really tell us much about what happened to the others.

Indeed removing the transactions eagerly can be problematic (I've outlined one use case in the docs) - but the idea is that caller of that function may re-verify the transactions that got pruned and reimport them to the pool if they are still valid.

We prune the transactions eagerly for two reasons:
1. If we possibly miss a block notification we will get the pool into consistent state
2. If we don't do that we can easily end up with plenty stale transactions in the pool that will never really be included.
So since the pool cannot re-verify the transactions itself, I chose to remove potentially more than needed, but allowing the caller to reimport transactions if he figures they are still valid.

Reply:
1. not sure why this would be the case.
2. i don't see why we would end up with stale transactions - if they *can* be included (and by definition, D and E still can, even when C has been included) then they should be in the queue.

Reply:
Block notifications are spuriously missed, not delivered during "major sync" (can restart at any time if we go offline for a little, and are generally unreliable

Reply:
If `C` was included (and it required `B` and `D`) you cannot really include `D` or `D'` providing the same tag, as the assumption is that every tag can be provided by only one transaction.
Also as mentioned, this code will be wrapped in a higher-level struct that will attempt to revalidate all transactions to make sure if they are still OK and import them back if that's the case.